runtime/calamari/src/zenlink.rs

```diff
@@ -0,0 +1,283 @@
+    pub SelfParaId: u32 = ParachainInfo::parachain_id().into();
+    pub MantaNativeAssetId: CalamariAssetId = 1;
+    pub ZenlinkNativeAssetId: u64 = 0;
+
+    pub const AnyNetwork: NetworkId = NetworkId::Any;
+    // Not allowed parachain token transfer on zenlink protocol
+    pub ZenlinkRegistedParaChains: Vec<(MultiLocation, u128)> = vec![];
+}
+
+type MultiAssets = ZenlinkMultiAssets<ZenlinkProtocol, Balances, LocalAssetAdaptor>;
+
+pub type ZenlinkLocationToAccountId = (
+    // Sibling parachain origins convert to AccountId via the `ParaId::into`.
+    SiblingParachainConvertsVia<Sibling, AccountId>,
+    // Straight up local `AccountId32` origins just alias directly to `AccountId`.
+    AccountId32Aliases<RelayNetwork, AccountId>,
+);
+
+impl zenlink_protocol::Config for Runtime {
+    type RuntimeEvent = RuntimeEvent;
+    type MultiAssetsHandler = MultiAssets;
+    type PalletId = ZenlinkPalletId;
+    type SelfParaId = SelfParaId;
+    type TargetChains = ZenlinkRegistedParaChains;
+    type AssetId = ZenlinkAssetId;
+    type AssetIdConverter = MantaAssetIdConverter;
+    #[cfg(not(feature = "runtime-benchmarks"))]
+    type LpGenerate = AssetManagerLpGenerate;
+    #[cfg(feature = "runtime-benchmarks")]
+    type LpGenerate = mock_benchmark::MockAssetManagerLpGenerate;
+    type AccountIdConverter = ZenlinkLocationToAccountId;
+    type XcmExecutor = ();
+    type WeightInfo = ();
+}
+
+pub struct MantaAssetIdConverter;
+impl ConvertMultiLocation<ZenlinkAssetId> for MantaAssetIdConverter {
+    fn chain_id(asset_id: &ZenlinkAssetId) -> u32 {
+        asset_id.chain_id
+    }
+
+    fn make_x3_location(asset_id: &ZenlinkAssetId) -> MultiLocation {
+        let asset_index = asset_id.asset_index;
+        if asset_index == ZenlinkNativeAssetId::get() {
+            // Notice: native asset is register as (1, Parachain(id)) location now.
+            return MultiLocation::new(1, X1(Parachain(SelfParaId::get())));
+        }
+        let asset = asset_index as CalamariAssetId;
+        let asset_location: AssetLocation =
+            AssetManager::location(&asset).expect("Asset should have Location!");
+        asset_location
+            .0
+            .try_into()
+            .expect("Location convert should not failed!")
+    }
+}
+
+pub struct AssetManagerLpGenerate;
+impl GenerateLpAssetId<ZenlinkAssetId> for AssetManagerLpGenerate {
+    fn generate_lp_asset_id(
+        asset_0: ZenlinkAssetId,
+        asset_1: ZenlinkAssetId,
+    ) -> Option<ZenlinkAssetId> {
+        // LP asset id is registered on AssetManager based on two asset id.
+        let asset_id_0 = LocalAssetAdaptor::asset_id_convert(asset_0);
+        let asset_id_1 = LocalAssetAdaptor::asset_id_convert(asset_1);
+        if asset_id_0.is_none() || asset_id_1.is_none() {
+            return None;
+        }
+        let lp_asset_id: Option<CalamariAssetId> =
+            <AssetManager as AssetIdLpMap>::lp_asset_id(&asset_id_0.unwrap(), &asset_id_1.unwrap());
```

Comment:
unwraps here? shouldn't we return none instead

Reply:
Oh already is checked above. Probably should put expects() instead with qed. Or we could do a match statement instead